### DISCUSSION TITLE
[Question] Multistage sequencer

### DISCUSSION THREAD

**Discussion by Garrett-C** [tsunamayo/Starship-EVO#3945] [Suggestions]
With the new novox tech allowing people to build larger ships without harming performance as much, I was recently wondering how people and NPCs could traverse these larger ships and the two options that came to me were a monorail esc system for horizontal movement and elevators for vertical. As things stand building these methods of movement for more than 2 points requires a lot of mechanisms and a lot of logic. Which is fine but not the most intuitive or player-friendly. 

Now in the past, you could connect many sequencers to one system which allowed this function to work well and this is certainly one option for allowing the creation of these systems. That said I do agree that the sequencer acting this way does seem odd as it would be the only element of logic to behave like this. So perhaps a new specialised sequencer could be the solution. Maybe it allows the user to have multiple states. Now obviously the issue this presents is that it's tough to control a sequencer like this. On/off is an easy way to control a sequencer but one with say 10 states between is not as easy. I don't have an elegant solution to this but perhaps there could be another element added to logic, like channels for example. 

Channels could function by passing your on/off signal to a channel logic block, the channel can be configured in here. Then the output from the channel system would be an on/off signal with an assigned channel. This could then by passed to the advanced sequencer which has different states for different channels. 

Something like this:
![image](https://user-images.githubusercontent.com/43644633/122625974-c8ab9e80-d09f-11eb-8c7a-fe9a63676e9e.png)

I don't know that this is the best solution and am sure you have some ideas cooking but figured I would throw some ideas out there to see what the thoughts on this area are.

**Comment by tsunamayo** [comment]
Yep it is a major gap I need to address. I need to spend more time thinking about it, I also want to introduce a "Push config override" gate that could achieve this. Thx

**Reply by Garrett-C** [reply]
Oh that does sound like a good way of doing it yeah. Plus an overide gate sounds like it could have loads of uses outside of just elevators and trams.

**Comment by Uncle-Ulty** [comment]
I have an idea, and idk if it's hard to implement. Instead of adding several different logic blocks for channels and switches, would it be possible to add an "interface switch controller"? It would be a block that you press "f" and it opens a window with all the multichannel / multistage options, and you press the desired option. So, instead of adding 10 different switches, add just only one interface switch, which would open a customizable pop-up window with available options.

This system would require two new blocks, a multistage sequencer with multichannel input, and a multichannel switch controller.

the sequencer would have an adjustable number of sequencer output for each channel (2,3,4,5 and so on), each one for a specific #channel.

and the interface switch controller would have the same adjustable number of channels and for each channel a small text space to specify what is it (elevator floors, or tram stations, or % load on a specific block such as engine or shield and so on)

I know this struggle, I made a ship with a 4-floor elevator and a 3 stations tram. It requires a lot of logic and time to make it work!

**Reply by Garrett-C** [reply]
I think that's sort of what I was trying to suggest maybe.

My suggestion was to have a block that sort of takes a generic input and translates it to a channel. Then the sequencer could then understand these different channels. So for each channel you want to use you have a channel encoder or whatever it could be called. This solves the issue in a ui of knowing which input is which. I think this is sort of similar to what you are saying here. 

That said I do sort of feel like the whole channel thing is a dodgy way of doing it because it sort of overcomplicates logic to an extent. Would be nice if there was a way of doing it without making logic much messier. Like the override block tsuna seems to be considering. 

But yeah I have a lot of floors to try to add to this ship so any better system would be ideal lol.

**Reply by Uncle-Ulty** [reply]
> I think that's sort of what I was trying to suggest maybe.
> 
> My suggestion was to have a block that sort of takes a generic input and translates it to a channel. Then the sequencer could then understand these different channels. So for each channel you want to use you have a channel encoder or whatever it could be called. This solves the issue in a ui of knowing which input is which. I think this is sort of similar to what you are saying here.
> 
> That said I do sort of feel like the whole channel thing is a dodgy way of doing it because it sort of overcomplicates logic to an extent. Would be nice if there was a way of doing it without making logic much messier. Like the override block tsuna seems to be considering.
> 
> But yeah I have a lot of floors to try to add to this ship so any better system would be ideal lol.

I got your idea. But what would happen with 2 or more channels were in "on" mode?

**Reply by Garrett-C** [reply]
Good question.  I don't think it overly matters as long as it's consistent. My suggestion would be that the channel with the lower number assumes priority. So if channel 1 was linked to say 5 meters and channel 2 was linked to say 10 meters if both were on the rail would go to 5m. 

That said this very question is a good indicator as to why channels shouldn't be the solution to the issue. Way too overcomplicated.

**Reply by Uncle-Ulty** [reply]
yeah, unless de sequencer only accepts the last input as the priority. I think we should work on both ways. a new channel logic block, and a multichannel interface controller, for cases such as elevator and tram.
